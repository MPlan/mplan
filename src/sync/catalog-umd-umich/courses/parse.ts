// import { Course } from 'sync/catalog-umd-umich/models';
import { JSDOM } from 'jsdom';
import { parseCourseBlockTitle } from './parse-course-block-title';
import { parseCourseBlockExtra } from './parse-course-block-extra';
import { Prerequisite, Course } from '../models';

export function parseCourses(courseListHtml: string): Course[] {
  const { window } = new JSDOM(courseListHtml);
  const { document } = window;
  const courseBlocks = Array.from(document.querySelectorAll('.courseblock'));
  return courseBlocks.map(courseBlock => {
    // subject code, course number, name, and credit hours
    const courseBlockTitleEl = courseBlock.querySelector('.courseblocktitle');
    if (!courseBlockTitleEl) throw new Error('could not find .courseblocktitle');
    const courseBlockTitle = courseBlockTitleEl.textContent;
    if (!courseBlockTitle) throw new Error('could not get text content of .courseblocktitle');
    const { courseNumber, creditHours, name, subjectCode } = parseCourseBlockTitle(
      courseBlockTitle,
    );

    // description
    const descriptionEl = courseBlock.querySelector('.courseblockdesc');
    if (!descriptionEl) {
      throw new Error('could not get description because could not find .courseblockdesc');
    }
    const _description = descriptionEl.textContent;
    if (!_description) throw new Error('description block text content was falsy');
    const description = _description.trim();

    const extras = Array.from(courseBlock.querySelectorAll('.courseblockextra'))
      .map(block => {
        try {
          return parseCourseBlockExtra(block);
        } catch (e) {
          console.warn(`Couldn't parse ${subjectCode} ${courseNumber}`, e);
          return undefined;
        }
      })
      .filter(x => !!x)
      .reduce(
        (extras, next) => {
          return { ...extras, ...next };
        },
        {} as any,
      ) as { prerequisites?: Prerequisite; corequisites?: Prerequisite; restrictions?: string[] };

    return {
      subjectCode,
      courseNumber,
      name,
      creditHours,
      description,
      ...extras,
    };
  });
}
