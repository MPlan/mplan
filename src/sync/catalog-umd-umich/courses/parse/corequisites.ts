export function parseCorequisites(courseBlockExtra: Element) {
  
}