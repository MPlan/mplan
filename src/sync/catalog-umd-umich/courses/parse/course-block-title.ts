import { oneLine } from 'common-tags';

export function parseCourseBlockTitle(courseBlockTitle: string) {
  const subjectCodeCourseNumberMatch = /(\w*)\s*(\w*)/.exec(courseBlockTitle);
  if (!subjectCodeCourseNumberMatch) {
    throw new Error(
      oneLine`
        could not get the subject code and course number from  course block title
        "${courseBlockTitle}"
      `,
    );
  }

  const subjectCode = subjectCodeCourseNumberMatch[1].toUpperCase().trim();
  const courseNumber = subjectCodeCourseNumberMatch[2].toUpperCase().trim();
  const nameAndCreditHours = courseBlockTitle.substring(
    subjectCodeCourseNumberMatch.index + subjectCodeCourseNumberMatch[0].length,
    courseBlockTitle.length,
  );

  const creditHourRangeMatch = /(\d)+\s*(?:to|or)\s*(\d+)\s*credit\s*hours?/i.exec(
    nameAndCreditHours,
  );
  if (creditHourRangeMatch) {
    const creditHourMin = parseInt(creditHourRangeMatch[1].trim(), 10);
    const creditHourMax = parseInt(creditHourRangeMatch[2].trim(), 10);

    const name = nameAndCreditHours.substring(0, creditHourRangeMatch.index).trim();
    return {
      subjectCode,
      courseNumber,
      name,
      creditHours: [creditHourMin, creditHourMax] as [number, number],
    };
  }

  const creditHourMatch = /(\d+)\s*credit\s*hours?/i.exec(nameAndCreditHours);
  if (creditHourMatch) {
    const creditHours = parseInt(creditHourMatch[1].trim(), 10);

    const name = nameAndCreditHours.substring(0, creditHourMatch.index).trim();
    return {
      subjectCode,
      courseNumber,
      name,
      creditHours,
    };
  }
  throw new Error(
    oneLine`
      could not parse name andd credit hours from course block title "${courseBlockTitle}"
    `,
  );
}
