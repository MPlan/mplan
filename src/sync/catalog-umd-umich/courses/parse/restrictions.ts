export function parseRestrictionsBlock(restrictionsBlock: Element) {
  return Array.from(restrictionsBlock.childNodes)
    .map(x => x.textContent!)
    .filter(x => !!x)
    .map(x => x.trim())
    .filter(x => !!x)
    .filter(x => !/restriction/i.test(x));
}
