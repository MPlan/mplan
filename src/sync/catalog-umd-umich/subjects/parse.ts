import { Subject } from 'sync/catalog-umd-umich/models';
import { JSDOM } from 'jsdom';

export function parseTextContent(textContent: string) {
  const match = /(.*)\((.*)\)/.exec(textContent);
  if (!match) throw new Error(`Could not parse text content of "${textContent}"`);
  const name = match[1].trim();
  const code = match[2].trim();
  return { name, code };
}

export function parseSubjects(html: string): Subject[] {
  const { window } = new JSDOM(html);
  const { document } = window;
  const aToZIndex = document.querySelector('#atozindex');
  if (!aToZIndex) throw new Error('could not find `#atozinddex` in the document');
  const anchors = Array.from(aToZIndex.querySelectorAll('ul li a')) as HTMLAnchorElement[];

  return anchors.map(anchor => ({
    href: anchor.href,
    ...parseTextContent(anchor.textContent || ''),
  }));
}
