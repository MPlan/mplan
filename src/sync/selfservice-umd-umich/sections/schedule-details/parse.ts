import { JSDOM } from 'jsdom';
import { formDecode } from 'utilities/form-encode-decode';
import { regularToCamelCase } from 'utilities/regular-to-camel-case';
import { range } from 'lodash';
import { decode } from 'he';

export function parseCapacityAndRemaining(seatsTbody: HTMLTableSectionElement) {
  const empty = { capacity: NaN, remaining: NaN };
  const rows = Array.from(seatsTbody.children).filter(
    elem => elem.tagName.toLowerCase().trim() === 'tr',
  );

  const headerRow = rows[0];

  if (!headerRow) {
    return empty;
  }

  const headings = Array.from(headerRow.children)
    .slice(1)
    .map(cell => cell.textContent || '')
    .map(text => regularToCamelCase(text));

  const result = rows.slice(1).reduce(
    (obj, row) => {
      const cells = Array.from(row.children);
      const firstCell = cells[0];
      if (!firstCell) {
        return obj;
      } // continue

      const key = regularToCamelCase(firstCell.textContent || '').toLowerCase();
      const values = cells
        .slice(1)
        .map(cell => cell.textContent || '')
        .map(text => parseInt(text.trim()));

      const value = range(values.length)
        .map(i => ({ key: headings[i], value: values[i] }))
        .reduce(
          (obj, { key, value }) => {
            obj[key] = value;
            return obj;
          },
          {} as { [key: string]: number },
        );

      obj[key] = value;

      return obj;
    },
    {} as { [seatType: string]: { [heading: string]: number } },
  );

  if (result.crosslistseats) {
    const capacity = result.crosslistseats && result.crosslistseats.capacity;
    const remaining = result.crosslistseats && result.crosslistseats.remaining;
    return { capacity, remaining };
  } else if (result.seats) {
    const capacity = result.seats && result.seats.capacity;
    const remaining = result.seats && result.seats.remaining;
    return { capacity, remaining };
  }

  return empty;
}

export function parseCrossListedCourses(infoCell: Element) {
  const innerHtml = infoCell.innerHTML;
  const matchWithPrerequisites = /cross\s*list\s*courses([\s\S]*)(?:prerequisites)/i.exec(
    innerHtml,
  );
  const matchWithoutPrerequisites = /cross\s*list\s*courses([\s\S]*)/i.exec(infoCell.innerHTML);
  const match = matchWithPrerequisites || matchWithoutPrerequisites;
  if (!match) {
    return [];
  }

  try {
    const document = new JSDOM(match[1]).window.document;
    const anchors = Array.from(document.querySelectorAll('a'));

    const crossListedCourses = anchors
      .map(anchor => anchor.href)
      .filter(x => !!x)
      .map(href => {
        const match = /\?(.*)/.exec(href);
        if (!match) {
          return undefined;
        }
        const decoded = formDecode(decode(match[1]));
        return [
          decoded.one_subj.trim().toUpperCase(),
          decoded.sel_crse_strt.trim().toUpperCase(),
        ] as [string, string];
      })
      .filter(x => x && x.length == 2)
      .map(x => x!);

    return crossListedCourses;
  } catch (e) {
    console.warn('Error in parsing cross listed courses.', e);
    return [];
  }
}

/**
 * captures the `capacity` and `remaining` of a section detail HTML
 */
export function parseScheduleDetail(html: string, logger?: (message: string) => void) {
  const log = logger || console.warn.bind(console);
  try {
    const document = new JSDOM(html).window.document;

    const seatsTbody = (document.querySelector('.datadisplaytable tbody .datadisplaytable tbody') ||
      document.createElement('tbody')) as HTMLTableSectionElement;

    const infoCell =
      document.querySelector('.datadisplaytable tbody .dddefault') || document.createElement('td');

    const capAndRem = parseCapacityAndRemaining(seatsTbody);
    const crossList = parseCrossListedCourses(infoCell);

    return {
      ...capAndRem,
      crossList,
    };
  } catch (e) {
    log(`Failed to parse a schedule detail: ${e && e.message}`);
    return undefined;
  }
}
