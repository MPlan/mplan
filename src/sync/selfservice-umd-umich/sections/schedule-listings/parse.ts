import { JSDOM } from 'jsdom';
import { isEqual } from 'lodash';
import { formDecode } from 'utilities/form-encode-decode';
import { regularToCamelCase } from 'utilities/regular-to-camel-case';

interface SectionTableParseResult {
  instructors: string[];
  times: string[];
  days: string[];
  locations: string[];
}

export function parseScheduleListing(html: string, logger?: (message: string) => void) {
  const log = logger || console.warn.bind(console);
  const document = new JSDOM(html).window.document;

  const dataDisplayTable = document.querySelector('.pagebodydiv > .datadisplaytable');
  if (!dataDisplayTable) {
    return [];
  }

  const tableBody = dataDisplayTable.querySelector('tbody');
  if (!tableBody) {
    return [];
  }

  const tableRows = Array.from(tableBody.children).filter(
    child => child.tagName === 'TR',
  ) as HTMLTableRowElement[];

  const sectionGroups = tableRows.reduce(
    (arr, headerOrBody, index) => {
      const arrIndex = Math.floor(index / 2);
      arr[arrIndex] = arr[arrIndex] || {};
      if (index % 2 === 0) {
        arr[arrIndex].header = headerOrBody;
      } else {
        arr[arrIndex].body = headerOrBody;
      }
      return arr;
    },
    [] as Array<{ header: HTMLTableRowElement; body: HTMLTableRowElement }>,
  );

  const result = sectionGroups
    .map(group => {
      const anchorWithCrn = group.header.querySelector('a');
      const crn = (anchorWithCrn && formDecode(anchorWithCrn.href || '').crn_in) || '';
      try {
        const section = parseSectionElement(group.body);
        return { crn, section };
      } catch (e) {
        log(`Could not parse scheduling list with CRN ${crn}: ${e && e.message}`);
        return { crn, section: (undefined as any) as SectionTableParseResult };
      }
    })
    .filter(({ crn, section }) => {
      if (!crn) return false;
      if (!section) return false;
      return true;
    })
    .map(({ crn, section }) => ({
      courseRegistrationNumber: crn,
      ...section,
    }));

  return result;
}

function findUniqueName(element: Element) {
  const anchor = element.querySelector('a');
  if (!anchor) {
    return undefined;
  }
  const href = anchor.href;
  if (!href) {
    return undefined;
  }
  const match = /mailto:(.*)@umich\.edu/i.exec(href);
  if (!match) {
    return undefined;
  }
  return match[1].trim().toLowerCase();
}

const emptySectionResult: SectionTableParseResult = {
  days: [],
  instructors: [],
  locations: [],
  times: [],
};

function parseSectionElement(body: HTMLTableRowElement | undefined | null) {
  if (!body) {
    return emptySectionResult;
  }

  const sectionTBody = body.querySelector('.datadisplaytable > tbody');
  if (!sectionTBody) {
    return emptySectionResult;
  }

  const rows = Array.from(sectionTBody.children).filter(
    child => child.tagName.toLowerCase().trim() === 'tr',
  ) as HTMLTableRowElement[];

  const headerRow = rows[0];
  if (!headerRow) {
    return emptySectionResult;
  }

  const headers = Array.from(headerRow.children)
    .filter(child => child.classList.contains('ddheader'))
    .map(child => child.textContent || '')
    .map(text => regularToCamelCase(text));

  const result = rows.slice(1).reduce(
    (obj, nextRow) => {
      const cells = Array.from(nextRow.querySelectorAll('.dddefault'))
        .map(cell => findUniqueName(cell) || cell.textContent || '')
        .map(text => text.trim());

      for (let i = 0; i < cells.length; i += 1) {
        const next = cells[i];
        const current = obj[headers[i]] || [];
        if (isEqual([next], current)) {
          continue;
        }
        const arr = current;
        obj[headers[i]] = [...arr, next];
      }

      return obj;
    },
    {} as { [key: string]: undefined | string[] },
  );

  const section: SectionTableParseResult = {
    instructors: result.instructors || [],
    times: result.time || [],
    days: result.days || [],
    locations: result.where || [],
  };

  return section;
}
