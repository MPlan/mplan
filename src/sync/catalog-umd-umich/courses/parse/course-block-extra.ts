import { parsePrerequisites } from './prerequisites';
import { parseRestrictionsBlock } from './restrictions';

export function parseCourseBlockExtra(courseBlockExtra: Element) {
  const textContent = courseBlockExtra.textContent;
  if (!textContent) throw new Error('could not get textContent from .courseblockextra');
  if (/restriction/i.test(textContent)) {
    return { restrictions: parseRestrictionsBlock(courseBlockExtra) };
  }
  if (/prerequisite/i.test(textContent)) {
    return { prerequisites: parsePrerequisites(courseBlockExtra) };
  }
  if (/corequisite/i.test(textContent)) {
    return { corequisites: parsePrerequisites(courseBlockExtra) };
  }
  return undefined;
}
