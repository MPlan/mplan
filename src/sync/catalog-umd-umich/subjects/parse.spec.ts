import { getExampleHtml } from 'testing-utilities/example-html';
import { parseSubjects } from './parse';

describe('subjects', () => {
  describe('parseSubjects', () => {
    it('parses all the subjects from the HTML page', () => {
      const html = getExampleHtml('undergrad-subjects');
      const subjects = parseSubjects(html);
      expect(subjects).toMatchSnapshot();
    });
  });
});
