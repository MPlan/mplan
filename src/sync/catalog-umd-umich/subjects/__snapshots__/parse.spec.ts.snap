// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`subjects parseSubjects parses all the subjects from the HTML page 1`] = `
Array [
  Object {
    "code": "ACC",
    "href": "/undergraduate/coursesaz/acc/",
    "name": "Accounting",
  },
  Object {
    "code": "AAAS",
    "href": "/undergraduate/coursesaz/aaas/",
    "name": "African & African-Amer Studies",
  },
  Object {
    "code": "AMST",
    "href": "/undergraduate/coursesaz/amst/",
    "name": "American Studies",
  },
  Object {
    "code": "ANTH",
    "href": "/undergraduate/coursesaz/anth/",
    "name": "Anthropology",
  },
  Object {
    "code": "MAPP",
    "href": "/undergraduate/coursesaz/mapp/",
    "name": "Applied Music",
  },
  Object {
    "code": "AAST",
    "href": "/undergraduate/coursesaz/aast/",
    "name": "Arab American Studies",
  },
  Object {
    "code": "ARBC",
    "href": "/undergraduate/coursesaz/arbc/",
    "name": "Arabic",
  },
  Object {
    "code": "ART",
    "href": "/undergraduate/coursesaz/art/",
    "name": "Art Applied",
  },
  Object {
    "code": "ARTH",
    "href": "/undergraduate/coursesaz/arth/",
    "name": "Art History",
  },
  Object {
    "code": "ASTR",
    "href": "/undergraduate/coursesaz/astr/",
    "name": "Astronomy",
  },
  Object {
    "code": "BBS",
    "href": "/undergraduate/coursesaz/bbs/",
    "name": "Behavioral and Biological Sciences",
  },
  Object {
    "code": "BCHM",
    "href": "/undergraduate/coursesaz/bchm/",
    "name": "Biochemistry",
  },
  Object {
    "code": "BENG",
    "href": "/undergraduate/coursesaz/beng/",
    "name": "Bioengineering",
  },
  Object {
    "code": "BIOL",
    "href": "/undergraduate/coursesaz/biol/",
    "name": "Biological Science",
  },
  Object {
    "code": "BA",
    "href": "/undergraduate/coursesaz/ba/",
    "name": "Business Administration",
  },
  Object {
    "code": "BE",
    "href": "/undergraduate/coursesaz/be/",
    "name": "Business Economics",
  },
  Object {
    "code": "BI",
    "href": "/undergraduate/coursesaz/bi/",
    "name": "Business Internship",
  },
  Object {
    "code": "BPS",
    "href": "/undergraduate/coursesaz/bps/",
    "name": "Business Policy and Strategy",
  },
  Object {
    "code": "CHEM",
    "href": "/undergraduate/coursesaz/chem/",
    "name": "Chemistry",
  },
  Object {
    "code": "CLS",
    "href": "/undergraduate/coursesaz/cls/",
    "name": "Child Life",
  },
  Object {
    "code": "CIVE",
    "href": "/undergraduate/coursesaz/cive/",
    "name": "Civic Engagement",
  },
  Object {
    "code": "COMM",
    "href": "/undergraduate/coursesaz/comm/",
    "name": "Communication",
  },
  Object {
    "code": "CHE",
    "href": "/undergraduate/coursesaz/che/",
    "name": "Community Health Education",
  },
  Object {
    "code": "COML",
    "href": "/undergraduate/coursesaz/coml/",
    "name": "Comparative Literature",
  },
  Object {
    "code": "CCM",
    "href": "/undergraduate/coursesaz/ccm/",
    "name": "Computer & Computational Math",
  },
  Object {
    "code": "CIS",
    "href": "/undergraduate/coursesaz/cis/",
    "name": "Computer & Information Science",
  },
  Object {
    "code": "CRJ",
    "href": "/undergraduate/coursesaz/crj/",
    "name": "Criminal Justice Studies",
  },
  Object {
    "code": "DS",
    "href": "/undergraduate/coursesaz/ds/",
    "name": "Decision Sciences",
  },
  Object {
    "code": "ECON",
    "href": "/undergraduate/coursesaz/econ/",
    "name": "Economics",
  },
  Object {
    "code": "EDA",
    "href": "/undergraduate/coursesaz/eda/",
    "name": "Educ A-Theoretical Foundatns",
  },
  Object {
    "code": "EDB",
    "href": "/undergraduate/coursesaz/edb/",
    "name": "Educ B-Educational Admin",
  },
  Object {
    "code": "EDC",
    "href": "/undergraduate/coursesaz/edc/",
    "name": "Educ C-Psychological Foundatns",
  },
  Object {
    "code": "EDD",
    "href": "/undergraduate/coursesaz/edd/",
    "name": "Educ D-Curriculum & Instructn",
  },
  Object {
    "code": "EDF",
    "href": "/undergraduate/coursesaz/edf/",
    "name": "Educ F-Physical Education",
  },
  Object {
    "code": "EDK",
    "href": "/undergraduate/coursesaz/edk/",
    "name": "Educ K-Independent Study",
  },
  Object {
    "code": "EDM",
    "href": "/undergraduate/coursesaz/edm/",
    "name": "Educ M-Community & Bilingual",
  },
  Object {
    "code": "EDN",
    "href": "/undergraduate/coursesaz/edn/",
    "name": "Educ N-Special Education",
  },
  Object {
    "code": "EDT",
    "href": "/undergraduate/coursesaz/edt/",
    "name": "Educ T-Education Technology",
  },
  Object {
    "code": "ECE",
    "href": "/undergraduate/coursesaz/ece/",
    "name": "Electrical&Computer Engin",
  },
  Object {
    "code": "ENGR",
    "href": "/undergraduate/coursesaz/engr/",
    "name": "Engineering Core",
  },
  Object {
    "code": "COMP",
    "href": "/undergraduate/coursesaz/comp/",
    "name": "English Composition",
  },
  Object {
    "code": "ENGL",
    "href": "/undergraduate/coursesaz/engl/",
    "name": "English",
  },
  Object {
    "code": "ENT",
    "href": "/undergraduate/coursesaz/ent/",
    "name": "Entrepreneurship",
  },
  Object {
    "code": "ESCI",
    "href": "/undergraduate/coursesaz/esci/",
    "name": "Environmental Science",
  },
  Object {
    "code": "ENST",
    "href": "/undergraduate/coursesaz/enst/",
    "name": "Environmental Studies",
  },
  Object {
    "code": "EXPS",
    "href": "/undergraduate/coursesaz/exps/",
    "name": "Exploratory Studies",
  },
  Object {
    "code": "FIN",
    "href": "/undergraduate/coursesaz/fin/",
    "name": "Finance",
  },
  Object {
    "code": "FREN",
    "href": "/undergraduate/coursesaz/fren/",
    "name": "French",
  },
  Object {
    "code": "GEOG",
    "href": "/undergraduate/coursesaz/geog/",
    "name": "Geography",
  },
  Object {
    "code": "GEOL",
    "href": "/undergraduate/coursesaz/geol/",
    "name": "Geology",
  },
  Object {
    "code": "GER",
    "href": "/undergraduate/coursesaz/ger/",
    "name": "German",
  },
  Object {
    "code": "GLOC",
    "href": "/undergraduate/coursesaz/gloc/",
    "name": "Global Cultures",
  },
  Object {
    "code": "HHS",
    "href": "/undergraduate/coursesaz/hhs/",
    "name": "Health and Human Service",
  },
  Object {
    "code": "HPS",
    "href": "/undergraduate/coursesaz/hps/",
    "name": "Health Policy Studies",
  },
  Object {
    "code": "HIST",
    "href": "/undergraduate/coursesaz/hist/",
    "name": "History",
  },
  Object {
    "code": "MHIS",
    "href": "/undergraduate/coursesaz/mhis/",
    "name": "History of Music",
  },
  Object {
    "code": "HUM",
    "href": "/undergraduate/coursesaz/hum/",
    "name": "Humanities",
  },
  Object {
    "code": "HRM",
    "href": "/undergraduate/coursesaz/hrm/",
    "name": "Human Resource Management",
  },
  Object {
    "code": "IMSE",
    "href": "/undergraduate/coursesaz/imse/",
    "name": "Indust & Manufac Sys Engin",
  },
  Object {
    "code": "ITM",
    "href": "/undergraduate/coursesaz/itm/",
    "name": "Information Technology Mgmt",
  },
  Object {
    "code": "IB",
    "href": "/undergraduate/coursesaz/ib/",
    "name": "International Business",
  },
  Object {
    "code": "JPN",
    "href": "/undergraduate/coursesaz/jpn/",
    "name": "Japanese",
  },
  Object {
    "code": "JASS",
    "href": "/undergraduate/coursesaz/jass/",
    "name": "Journalism and Screen Studies",
  },
  Object {
    "code": "LAT",
    "href": "/undergraduate/coursesaz/lat/",
    "name": "Latin",
  },
  Object {
    "code": "LE",
    "href": "/undergraduate/coursesaz/le/",
    "name": "Law & Environment",
  },
  Object {
    "code": "LIBS",
    "href": "/undergraduate/coursesaz/libs/",
    "name": "Liberal Studies",
  },
  Object {
    "code": "LIBR",
    "href": "/undergraduate/coursesaz/libr/",
    "name": "Library Science",
  },
  Object {
    "code": "LING",
    "href": "/undergraduate/coursesaz/ling/",
    "name": "Linguistics",
  },
  Object {
    "code": "MKT",
    "href": "/undergraduate/coursesaz/mkt/",
    "name": "Marketing",
  },
  Object {
    "code": "MATH",
    "href": "/undergraduate/coursesaz/math/",
    "name": "Mathematics",
  },
  Object {
    "code": "ME",
    "href": "/undergraduate/coursesaz/me/",
    "name": "Mechanical Engineering",
  },
  Object {
    "code": "MICR",
    "href": "/undergraduate/coursesaz/micr/",
    "name": "Microbiology",
  },
  Object {
    "code": "MILS",
    "href": "/undergraduate/coursesaz/mils/",
    "name": "Military Science",
  },
  Object {
    "code": "MCL",
    "href": "/undergraduate/coursesaz/mcl/",
    "name": "Modern & Classical Language",
  },
  Object {
    "code": "MTHY",
    "href": "/undergraduate/coursesaz/mthy/",
    "name": "Music Theory",
  },
  Object {
    "code": "NSCI",
    "href": "/undergraduate/coursesaz/nsci/",
    "name": "Natural Science",
  },
  Object {
    "code": "OM",
    "href": "/undergraduate/coursesaz/om/",
    "name": "Operations Management",
  },
  Object {
    "code": "OB",
    "href": "/undergraduate/coursesaz/ob/",
    "name": "Organizational Behavior",
  },
  Object {
    "code": "PHIL",
    "href": "/undergraduate/coursesaz/phil/",
    "name": "Philosophy",
  },
  Object {
    "code": "PHYS",
    "href": "/undergraduate/coursesaz/phys/",
    "name": "Physics",
  },
  Object {
    "code": "POL",
    "href": "/undergraduate/coursesaz/pol/",
    "name": "Political Science",
  },
  Object {
    "code": "PDED",
    "href": "/undergraduate/coursesaz/pded/",
    "name": "Professional Education",
  },
  Object {
    "code": "PSYC",
    "href": "/undergraduate/coursesaz/psyc/",
    "name": "Psychology",
  },
  Object {
    "code": "RELS",
    "href": "/undergraduate/coursesaz/rels/",
    "name": "Religious Studies",
  },
  Object {
    "code": "STS",
    "href": "/undergraduate/coursesaz/sts/",
    "name": "Science and Technology Studies",
  },
  Object {
    "code": "SSCI",
    "href": "/undergraduate/coursesaz/ssci/",
    "name": "Social Sciences",
  },
  Object {
    "code": "SWK",
    "href": "/undergraduate/coursesaz/swk/",
    "name": "Social Work",
  },
  Object {
    "code": "SOC",
    "href": "/undergraduate/coursesaz/soc/",
    "name": "Sociology",
  },
  Object {
    "code": "SPAN",
    "href": "/undergraduate/coursesaz/span/",
    "name": "Spanish",
  },
  Object {
    "code": "SPEE",
    "href": "/undergraduate/coursesaz/spee/",
    "name": "Speech",
  },
  Object {
    "code": "STAT",
    "href": "/undergraduate/coursesaz/stat/",
    "name": "Statistics",
  },
  Object {
    "code": "URS",
    "href": "/undergraduate/coursesaz/urs/",
    "name": "Urban and Regional Studies",
  },
  Object {
    "code": "WGST",
    "href": "/undergraduate/coursesaz/wgst/",
    "name": "Women's and Gender Studies",
  },
]
`;
